source/geometry_model/ellipsoidal_chunk.cc

```diff
@@ -40,6 +42,326 @@ namespace aspect
+       * This code may be freely used and modified for any purpose
+       * providing that this copyright notice is included with it.
+       * SoftSurfer makes no warranty for this code, and cannot be held
+       * liable for any real or imagined damage resulting from its use.
+       * Users of this code must verify correctness for their application.
+       */
+      int pointNo = pointList.size();
+      int    wn = 0;    // the  winding number counter
+      int   j=pointNo-1;
+
+
+      // loop through all edges of the polygon
+      for (int i=0; i<pointNo; i++)
+        {
+          // edge from V[i] to  V[i+1]
+          if (pointList[j][1] <= point[1])
+            {
+              // start y <= P.y
+              if (pointList[i][1]  > point[1])      // an upward crossing
+                if (( (pointList[i][0] - pointList[j][0]) * (point[1] - pointList[j][1])
+                      - (point[0] -  pointList[j][0]) * (pointList[i][1] - pointList[j][1]) ) > 0)
+                  {
+                    // P left of  edge
+                    ++wn;            // have  a valid up intersect
+                  }
+            }
+          else
+            {
+              // start y > P.y (no test needed)
+              if (pointList[i][1]  <= point[1])     // a downward crossing
+                if (( (pointList[i][0] - pointList[j][0]) * (point[1] - pointList[j][1])
+                      - (point[0] -  pointList[j][0]) * (pointList[i][1] - pointList[j][1]) ) < 0)
+                  {
+                    // P right of  edge
+                    --wn;            // have  a valid down intersect
+                  }
+            }
+          j=i;
+        }
+      return (wn != 0);
+    }
+
+    /**
+     * the EllipsoidalChunkTopography class
+     */
+    // constructor
+    template <int dim>
+    EllipsoidalChunk<dim>::EllipsoidalChunkTopography::EllipsoidalChunkTopography()
+    {}
+
+
+    template <int dim>
+    double
+    EllipsoidalChunk<dim>::EllipsoidalChunkTopography::value (const double lon,
+                                                              const double lat) const
+    {
+      double value = 0;
+
+      switch (topo_type)
+        {
+          case NO_TOPOGRAPHY:
+            return 0;
+            break;
+
+          case PRM_EXACT:
+            for (unsigned int i = 0; i < point_lists.size(); i++)
+              {
+                Point<2> p1(lon * 180/numbers::PI,lat * 180/numbers::PI);
+
+                if (In2dPolygon(p1, point_lists[i]))
+                  {
+                    value = topography_values[i];
+                  }
+              }
+            return value;
+            break;
+
+          case PRM_UNIFORM_GRID_INTERPOLATED:
+            if (topography_data != NULL)
+              return static_cast<Functions::InterpolatedUniformGridData<2>*>(topography_data)->value (Point<2>(lat * 180/numbers::PI,
+                     lon * 180/numbers::PI));
+            break;

```

Comment:
If you get to the `break` here, you has just encountered a situation that shouldn't happen. Best to just abort the program via `Assert (false, ExcInternalError());`. Alternatively, you could just assert that `topography_data` is non-`NULL`, and then avoid the `if`.